flink-tests/src/test/java/org/apache/flink/test/checkpointing/CheckpointFailureManagerITCase.java

```diff
@@ -82,18 +85,77 @@ public void testAsyncCheckpointFailureTriggerJobFailed() throws Exception {
         JobGraph jobGraph = StreamingJobGraphGenerator.createJobGraph(env.getStreamGraph());
         try {
             // assert that the job only execute checkpoint once and only failed once.
-            TestUtils.submitJobAndWaitForResult(
+            submitJobAndWaitForResult(
                     cluster.getClusterClient(), jobGraph, getClass().getClassLoader());
         } catch (JobExecutionException jobException) {
             Optional<FlinkRuntimeException> throwable =
-                    ExceptionUtils.findThrowable(jobException, FlinkRuntimeException.class);
-            Assert.assertTrue(throwable.isPresent());
-            Assert.assertEquals(
-                    CheckpointFailureManager.EXCEEDED_CHECKPOINT_TOLERABLE_FAILURE_MESSAGE,
-                    throwable.get().getMessage());
+                    findThrowable(jobException, FlinkRuntimeException.class);
+            assertTrue(throwable.isPresent());
+            assertEquals(
+                    EXCEEDED_CHECKPOINT_TOLERABLE_FAILURE_MESSAGE, throwable.get().getMessage());
         }
         // assert that the job only failed once.
-        Assert.assertEquals(1, StringGeneratingSourceFunction.INITIALIZE_TIMES.get());
+        assertEquals(1, StringGeneratingSourceFunction.INITIALIZE_TIMES.get());
+    }
+
+    @Test
+    public void testSourceFailureTriggerJobFailed() throws Exception {
+        // given: Environment with failed source and no restart strategy.
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.enableCheckpointing(2000L);
+        env.setRestartStrategy(RestartStrategies.noRestart());
+        env.addSource(new FailedSource()).addSink(new DiscardingSink<>());
+        JobGraph jobGraph = StreamingJobGraphGenerator.createJobGraph(env.getStreamGraph());
+        try {
+            // when: Execute job once.
+            submitJobAndWaitForResult(
+                    cluster.getClusterClient(), jobGraph, getClass().getClassLoader());
+        } catch (JobExecutionException jobException) {
+            Optional<RuntimeException> throwable =
+                    findThrowable(jobException, RuntimeException.class);
+
+            // then: Job failed with expected exception.
+            assertTrue(throwable.isPresent());
+            assertEquals(FailedSource.SOURCE_FAILED_MSG, throwable.get().getMessage());
```

Comment:
These assertions seem a bit fragile to me (what if Flink wraps an exception into `RuntimeException`?).
And they are not actually checking the production code, but the test itself: without the fix the test will time out; and without the exception thrown the job might have exited for some other reason.

But the setup is quite simple IMO, so I'd remove them.

WDYT?

Reply:
I think it makes sense what you said but the main target to check that the cluster fails only by expected reason. But anyway I think I will rework this place to avoid fragility